### PR TITLE
rbd: use go-ceph for getImageMirroringStatus

### LINKED ISSUE #2120: RBD: convert getImageMirroingStatus from RBD CLI to go-ceph
https://github.com/ceph/ceph-csi/blob/7cbad9305f71b70040509ed854f1d65c7c624d4f/internal/rbd/mirror.go#L131-L161 Need to re-written using go-ceph

### PR DESCRIPTION
**Pull request by yati1998**
Currently, getImageMirroringStatus() is using RBD CLI.
This commit converts RBD CLI to go-ceph API.

Fixes: #2120

Signed-off-by: Yati Padia <ypadia@redhat.com>

### CODE REVIEW

**Comment by pkalever** [comment]
It's good to capture some context about why this is Draft currently. Maybe use a task list to track any pending item's progress?

**Comment by yati1998** [comment]
> It's good to capture some context about why this is Draft currently. Maybe use a task list to track any pending item's progress?

Nothing much to list @pkalever .  I am stuck with something here. But will try to finish this soon.

File: internal/rbd/replicationcontrollerserver.go [hunk 1]
```diff
@@ -47,7 +48,7 @@ type imageMirroringState string
 const (
 	// If the state is up+unknown means the rbd-mirror daemon is
 	// running and the image is demoted on both the clusters.
-	upAndUnknown imageMirroringState = "up+unknown"
+	unknown imageMirroringState = "unknown"
```
Comment (nixpanic) [inline internal/rbd/replicationcontrollerserver.go:51]: the comment above still refers to "up+unknown"

File: internal/rbd/replicationcontrollerserver.go [hunk 2]
```diff
@@ -491,6 +492,7 @@ func (rs *ReplicationServer) DemoteVolume(ctx context.Context,
 // image is present, mirroring is enabled and the image is in demoted state.
 // If yes it will resync the image to correct the split-brain.
 // FIXME: reduce complexity.
+// nolint:gocyclo,cyclop // reduce complexity
```
Comment (nixpanic) [inline internal/rbd/replicationcontrollerserver.go]: yuck, consider this as a trigger to be required to correct this.

File: internal/rbd/replicationcontrollerserver.go [hunk 3]
```diff
@@ -570,24 +587,27 @@ func (rs *ReplicationServer) ResyncVolume(ctx context.Context,
...
-			if imageMirroringState(s.State) != upAndUnknown {
-				util.UsefulLog(
-					ctx,
-					"peer site name=%s, mirroring state=%s, description=%s and lastUpdate=%s",
-					s.SiteName,
-					s.State,
-					s.Description,
-					s.LastUpdate)
-				ready = false
+		for _, s := range mirrorStatus.SiteStatuses {
```
Comment (nixpanic) [inline internal/rbd/replicationcontrollerserver.go]: Consider moving this for loop to a helper function that returns the true/false you can assign to `ready`. This makes the function simpler and no need for the `// nolint` anymore

**Review by nixpanic — CHANGES_REQUESTED** [review]
just a nit, looks good otherwise

File: internal/rbd/replicationcontrollerserver.go (hunk 1 above)
Comment (yati1998) [reply internal/rbd/replicationcontrollerserver.go:51]: oops. Noted

File: internal/rbd/replicationcontrollerserver.go (hunk 3 above)
Comment (yati1998) [reply internal/rbd/replicationcontrollerserver.go]: sure.

File: internal/rbd/replicationcontrollerserver.go [hunk 4]
```diff
@@ -559,7 +561,22 @@ func (rs *ReplicationServer) ResyncVolume(ctx context.Context,
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	ready := false
-	state := imageMirroringState(mirrorStatus.State)
+
+	localStatus, err := mirrorStatus.LocalStatus()
+	if err != nil {
+		util.UsefulLog(
```
Comment (Madhu-1) [inline internal/rbd/replicationcontrollerserver.go]: incase of error the` localStatus` values be empty? do we need to log it?

File: internal/rbd/replicationcontrollerserver.go [hunk 5]
```diff
@@ -487,6 +488,29 @@ func (rs *ReplicationServer) DemoteVolume(ctx context.Context,
 	return &replication.DemoteVolumeResponse{}, nil
 }
 
+// This function checks the state of the remote cluster.
+// It returns true if the statue is up and unknown.
+func checkSiteStatus(ctx context.Context, mirrorStatus *librbd.GlobalMirrorImageStatus) bool {
```
Comment (Madhu-1) [inline internal/rbd/replicationcontrollerserver.go]: function comment should start with the function name
```
checkSiteStatus checks the state ...
```

Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go]: the function name can be renamed to checkRemoteSiteStatus?

Comment (nixpanic) [reply internal/rbd/replicationcontrollerserver.go]: Agreeing with this, these are the only nits that I would point out as well.

Comment (yati1998) [reply internal/rbd/replicationcontrollerserver.go]: done.

**Review by Madhu-1 — DISMISSED** [review]

File: internal/rbd/replicationcontrollerserver.go [hunk 6]
```diff
@@ -596,12 +615,6 @@ func (rs *ReplicationServer) ResyncVolume(ctx context.Context,
 		}
 	}
 
-	util.UsefulLog(
```
Comment (Madhu-1) [inline internal/rbd/replicationcontrollerserver.go:599]: can you please keep this one as it is and log the local details.

Comment (yati1998) [reply internal/rbd/replicationcontrollerserver.go:599]: sure

**Review by nixpanic — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Pull request merged**